### PR TITLE
Migrate JS package to hybrid esm/cjs module

### PR DESCRIPTION
**Pull request by aurelien-reeves**
# Description

Make the javascript package a hybrid esm/cjs module

# Motivation & context

This is the continuation of a work started a few weeks ago to make cucumber-js internals compatible with both CJS and ESM packages.

## Type of change

- Refactoring/debt (improvement to code design or tooling without changing behaviour)

## Note to contributors

I bring also two small addition to the PR:
- the use of `shx` to make life easier for non-POSIX (aka windows) contributors
- the addition of dedicated vscode eslint configuration. Per default vscode linter looks for some config in the root folder. As this is actually a monorepo, the dedicated configation tells vscode linter where to look for the javascript configuration

# Checklist:

- [x] I have read the [**CONTRIBUTING**](../CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [x] I have added tests to cover my changes. (`test:cjs` tests)
- [x] I have added an entry to the "Unreleased" section of the [**CHANGELOG**](../CHANGELOG.md), linking to this pull request.

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (aslakhellesoy) [inline .vscode/settings.json:1]: Was this file added intentionally?

File: javascript/package.json [hunk 2]
```diff
@@ -2,15 +2,31 @@
   "name": "@cucumber/ci-environment",
   "version": "8.0.1",
   "description": "Detect CI Environment from environment variables",
-  "main": "dist/src/index.js",
-  "types": "dist/src/index.d.ts",
+  "main": "dist/cjs/src/index.js",
+  "types": "dist/cjs/src/index.d.ts",
+  "files": [
+    "dist/cjs",
+    "dist/esp"
```
Comment (aslakhellesoy) [inline javascript/package.json:9]: Typo?

File: javascript/package.json [hunk 3]
```diff
@@ -40,6 +56,7 @@
     "glob": "7.2.0",
     "mocha": "9.1.3",
     "prettier": "2.5.1",
+    "shx": "^0.3.3",
```
Comment (aslakhellesoy) [inline javascript/package.json:59]: Let’s be consistent about pinned dependencies

File: javascript/src/detectCiEnvironment.ts [hunk 4]
```diff
@@ -1,3 +1,5 @@
+import { URL } from 'url'
```
Comment (aslakhellesoy) [inline javascript/src/detectCiEnvironment.ts:1]: Why is this imported? Node’s URL class is deprecated. We should use the global `URL` class.

**Review by aslakhellesoy — CHANGES_REQUESTED** [review]
Thanks for looking into this!

**Review by aslakhellesoy — COMMENTED** [review]
I’m surprised there are no failures caused by imports that reference directories or files without an explicit `.js` extension. I think something is missing…

File: .vscode/settings.json (hunk 1 above)
Comment (aslakhellesoy) [reply .vscode/settings.json:1]: Never mind - I see the explanation in the PR.

**Review by aslakhellesoy — CHANGES_REQUESTED** [review]
You’ve forgotten to add `"type": "module"` in `package.json`. Adding that will expose a few more things that need to be fixed, like imports.

File: javascript/package.json (hunk 2 above)
Comment (aurelien-reeves) [reply javascript/package.json:9]: Yeap 😅

File: javascript/src/detectCiEnvironment.ts (hunk 4 above)
Comment (aurelien-reeves) [reply javascript/src/detectCiEnvironment.ts:1]: No, Nod's `url` class is deprecated. This is the `URL` one which is recommended.
It is available globally, but typescript was not happy with this. It is possible to explicitly import it from the `url` module as described in the documentation: https://nodejs.org/dist/latest-v16.x/docs/api/url.html#new-urlinput-base

**Comment by aurelien-reeves** [comment]
> I’m surprised there are no failures caused by imports that reference directories or files without an explicit `.js` extension. I think something is missing…



> You’ve forgotten to add `"type": "module"` in `package.json`. Adding that will expose a few more things that need to be fixed, like imports.

Indeed, I was surprised too, but I was not able to find the issue myself. Thanks to spot it! Everything should be fixed now :)

**Pull request merged**